src/main/java/org/jfrog/hudson/generic/ArtifactoryGenericConfigurator.java

```diff
@@ -39,397 +39,434 @@
+							build.setResult(Result.FAILURE);
+							throw new IllegalArgumentException(error);
+						}
+						isFiltered = MultiConfigurationUtils.isfiltrated(build, getArtifactoryCombinationFilter());
+					}
+
+					if (!isFiltered) {
+						GenericArtifactsDeployer artifactsDeployer = new GenericArtifactsDeployer(build,
+								ArtifactoryGenericConfigurator.this, listener, preferredDeployer);
+						artifactsDeployer.deploy();
+
+						List<Artifact> deployedArtifacts = artifactsDeployer.getDeployedArtifacts();
+						if (deployBuildInfo) {
+							new GenericBuildInfoDeployer(ArtifactoryGenericConfigurator.this, client, build,
+									listener, deployedArtifacts, buildDependencies, publishedDependencies).deploy();
+							String buildName = BuildUniqueIdentifierHelper.getBuildNameConsiderOverride(ArtifactoryGenericConfigurator.this, build);
+							// add the result action (prefer always the same index)
+							build.getActions().add(0, new BuildInfoResultAction(getArtifactoryUrl(), build, buildName));
+							build.getActions().add(new UnifiedPromoteBuildAction(build, ArtifactoryGenericConfigurator.this));
+						}
+					}
+
+					return true;
+				} catch (Exception e) {
+					e.printStackTrace(listener.error(e.getMessage()));
+				} finally {
+					client.close();
+				}
+
+				// failed
+				build.setResult(Result.FAILURE);
+				return true;
+			}
+		};
+	}
+
+	private boolean isMultiConfProject(AbstractBuild build) {
+		return (build.getProject().getClass().equals(MatrixConfiguration.class));
+	}
+
+	@Override
+	public DescriptorImpl getDescriptor() {
+		return (DescriptorImpl) super.getDescriptor();
+	}
+
+	@Override
+	public boolean isOverridingDefaultResolver() {
+		return resolverCredentialsConfig != null && resolverCredentialsConfig.isCredentialsProvided();
+	}
+
+	@Override
+	public Credentials getOverridingResolverCredentials() {
+		return overridingResolverCredentials;
+	}
+
+
+	@Override
+	public CredentialsConfig getResolverCredentialsConfig() {
+		return resolverCredentialsConfig;
+	}
+
+	@Extension(optional = true)
+	public static class DescriptorImpl extends AbstractBuildWrapperDescriptor {
+		private static final String DISPLAY_NAME = "Generic-Artifactory Integration";
+		private static final String CONFIG_PREFIX = "generic";
+
+		public DescriptorImpl() {
+			super(ArtifactoryGenericConfigurator.class, DISPLAY_NAME, CONFIG_PREFIX);
+		}
+
+		@SuppressWarnings("unused")
+		@JavaScriptMethod
+		public RefreshServerResponse refreshFromArtifactory(String url, String credentialsId, String username, String password, boolean overrideCredentials) {
+			return super.refreshDeployersFromArtifactory(url, credentialsId, username, password, overrideCredentials, false);
+		}
+
+		@SuppressWarnings("unused")
+		@RequirePOST
+		public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item project) {
+			return PluginsUtils.fillPluginCredentials(project);
+		}
+
+		@Override
+		public boolean isJiraPluginEnabled() {
```

Comment:
This method is duplicated in the code.
I suggest moving it to *PluginsUtils* and making it static, and reusing it from both locations.

Reply:
Good suggestion. Will look into within a few days.

Thanks for coming back. Much appreciated.

Reply:
Alright, I attempted to refactor it based on your suggestion, but I've struggled to make this happen. I'm not experienced Java coder, however your proposal seems to be impossible to achieve in Java as other Classes which declares the same method, already extending some other classes and hence it is impossible to extend `PluginUtils` class. Although there are some new finding in the code.

My new observation is:
* that in this file: **`ArtifactoryGenericConfigurator.java`** where my newly added method `isJiraPluginEnabled()` is defined under this class `DescriptorImpl` which extends **`AbstractBuildWrapperDescriptor`**  the method  is actually defined in its super class `AbstractBuildWrapperDescriptor`. Thus this `@overwritten` method is truely a duplicate code, so thus I have removed it.
* however, in this file: **`ArtifactoryRedeployPublisher`**, the same method is also defined under the same class called `DescriptorImpl`, but this one extends `BuildStepDescriptor<Publisher>` which doesn't declare that method. 
But dunno how to inherit it from the proposed `PluginUtils` class. 
The only improvement I can see here would be to move that method for this scenario also into supper class.

I think it would goes really beyond this PR and someone who maintains the plugin might do this and other refactoring to make the code cleaner and nicer.